dashboard/app/controllers/activities_controller.rb

```diff
@@ -69,28 +109,35 @@ def milestone
                     client_state.lines
                   end
 
-    render json: milestone_response(script_level: @script_level,
+    render json: milestone_response(script_level: script_level,
                                     total_lines: total_lines,
-                                    trophy_updates: @trophy_updates,
+                                    trophy_updates: trophy_updates,
                                     solved?: solved,
-                                    level_source: @level_source.try(:hidden) ? nil : @level_source,
-                                    level_source_image: @level_source_image,
-                                    activity: @activity,
-                                    new_level_completed: @new_level_completed,
+                                    level_source: level_source.try(:hidden) ? nil : level_source,
+                                    level_source_image: level_source_image,
+                                    activity: activity,
+                                    new_level_completed: new_level_completed,
                                     share_failure: share_failure)
 
     slog(:tag => 'activity_finish',
-         :script_level_id => @script_level.try(:id),
-         :level_id => @level.id,
+         :script_level_id => script_level.try(:id),
+         :level_id => level.id,
          :user_agent => request.user_agent,
          :locale => locale) if solved
 
     # log this at the end so that server errors (which might be caused by invalid input) prevent logging
-    log_milestone(@level_source, params)
+    log_milestone(level_source, params, lines)
   end
 
   private
 
+  def nil_or_int(value)
```

Comment:
[`JSONValue#value`](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/lib/json_value.rb#L15) might be reusable here. (Even though it's called `JSONValue` it's more like a "String to Integer/Float/Boolean" utility method.)